File: Python/NeetCode_Graph_Valid_Tree.py
from collections import deque

def validTree(n : int, edges: List[List[int]]) -> bool:
    g = [ [] for _ in range(n) ]
    for [src, tgt] in edges:
        g[src].append(tgt)
        g[tgt].append(src)

    visited = set()
    q = deque([])
    q.append(0)
    while q:
        top = q.popleft()
        if top in visited:
            continue
        visited.add(top)
        for tgt in g[top]:
            q.append(tgt)
    return len(edges) == n - 1 if len(visited) == n else False

File: Cpp/NeetCode_Graph_Valid_Tree.cpp
#include <iostream>
#include <vector>
#include <deque>

int validTree(int n, vector<vector<int>>& edges) {
  if (edges.size() != n - 1) {
    return false;
  }
  vector<vector<int>> graph(n);
  for (vector<int>& pair : edges) {
    int src = pair[0], tgt = pair[1];
    graph[src].push_back(tgt);
    graph[tgt].push_back(src);
  }

  set<int> visited;
  deque<int> q;
  q.push_back(0);
  while (!q.empty()) {
    int src = q.pop_front();
    visited.insert(src);
    for (int next : graph[src]) {
      if (visited.find(next) == visited.end()) {
	q.push_back(next);
      }
    }
  }
  return visited.size() == n;
}
